src/include/denc.h

```diff
@@ -283,29 +283,17 @@ using underlying_type_t = typename underlying_type<T>::type;
 }
 
 template<class It>
-struct is_const_iterator
-  : std::conditional_t<std::is_const_v<std::remove_pointer_t<typename It::pointer>>,
-		       std::true_type,
-		       std::false_type>
-{};
-template<>
-struct is_const_iterator<size_t> : std::false_type {};
-template<>
-struct is_const_iterator<ceph::buffer::list::contiguous_appender> : std::false_type {
-  // appender is used for *changing* the buffer
-};
-template<class It>
-inline constexpr bool is_const_iterator_v = is_const_iterator<It>::value;
+concept is_const_iterator =
+std::is_const_v<std::remove_pointer_t<typename It::pointer>>;
```

Comment:
> what's so special about the `const` here? couldn't we read from a mutable `iterator` the same way we'd read from a `const_iterator`? maybe [std::input_iterator](https://en.cppreference.com/w/cpp/iterator/input_iterator) better expresses our requirements here

there are 3 cases, `buffer::ptr::const_iterator`, `size_t`, and `ceph::buffer::list::contiguous_appender`. the "const_iterator" here is `buffer::ptr::const_iterator`, but ironically, `buffer::ptr::const_iterator`, among other things, it is missing

- `std::input_iterator_tag`
- it is not `std::movable`
- it cannot be dereferenced

apart from here, we don't have any use case for the `input_iterator`.

> 
> the templates that use this concept also require a `get_pos_add()` member function - should our concept enforce that too?
> 
> which ceph types actually qualify for use here? if it's only `buffer::ptr::const_iterator`, maybe we don't need the templates for iterator type

yeah. just a single `std::same_as<>` would do the trick.

Reply:
thanks @tchaikov. i'm just brainstorming, so don't feel obligated to indulge me here - but i have a feeling we could benefit from making these act more like normal c++ iterators. i see that each `buffer::ptr::iterator` stores 4 pointers and a bool:
```c++
    class iterator_impl {
      const ptr *bp;     ///< parent ptr
      const char *start; ///< starting pointer into bp->c_str()
      const char *pos;   ///< pointer into bp->c_str()
      const char *end_ptr;   ///< pointer to bp->end_c_str()
      const bool deep;   ///< if true, do not allow shallow ptr copies
```
when it should only need a single pointer - `const char* pos` by itself is a fully-functional `std::contiguous_iterator`, and `get_pos_add()` could easily be defined in terms of that:
```c++
// return a pointer to the current position and advance the iterator accordingly
auto get_pos_add(std::size_t n, std::contiguous_iterator auto &pos)
{
  auto r = &*pos;
  std::advance(pos, n);
  return r;
}
```
this could potentially allow ceph to deserialize data from containers like array/string/string_view/vector/span. but it could also be a messy project, especially if it requires any changes to the librados API

Reply:
agreed. a free function is better also, IMHO. actually, i have a local branch implemented the required bits to make it an `input_iterator`. but in light of lacking potential users, i didn't create a PR for this.